src/components/tl/sharp/tl_sharp_coll.c

```diff
@@ -0,0 +1,249 @@
+        return UCC_ERR_NOT_SUPPORTED;
+    }
+
+    return UCC_OK;
+}
+
+static ucc_status_t
+ucc_tl_sharp_mem_register(ucc_tl_sharp_context_t *ctx, void *addr,
+                          size_t length, ucc_tl_sharp_reg_t **reg)
+{
+    ucc_rcache_region_t *rregion;
+    ucc_tl_sharp_rcache_region_t *region;
+    ucc_status_t status;
+    ucc_tl_sharp_reg_t *r;
+
+    if (length < ctx->cfg.reg_threshold) {
+        *reg = &ucc_tl_sharp_reg_null;
+        return UCC_OK;
+    }
+
+    if (ctx->rcache) {
+        status = ucc_rcache_get(ctx->rcache, (void *)addr, length,
+                                PROT_READ|PROT_WRITE, NULL,  &rregion);
+        if (status != UCC_OK) {
+            tl_error(ctx->super.super.lib, "ucc_rcache_get failed");
+            return UCC_ERR_INVALID_PARAM;
+        }
+        region = ucc_derived_of(rregion, ucc_tl_sharp_rcache_region_t);
+        *reg   = &region->reg;
+    } else {
+        r = ucc_malloc(sizeof(ucc_tl_sharp_reg_t), "sharp reg");
+        if (!r) {
+            tl_error(ctx->super.super.lib, "failed to allocate reg data");
+            return UCC_ERR_NO_MEMORY;
+        }
+
+        sharp_coll_reg_mr(ctx->sharp_context, addr, length, &r->mr);
+        *reg = r;
+    }
+
+    return UCC_OK;
+}
+
+static ucc_status_t
+ucc_tl_sharp_mem_deregister(ucc_tl_sharp_context_t *ctx,
+                            ucc_tl_sharp_reg_t *reg)
+{
+    ucc_tl_sharp_rcache_region_t *region;
+
+    if (reg == &ucc_tl_sharp_reg_null) {
+        return UCC_OK;
+    }
+
+    if (ctx->rcache) {
+        region = ucc_container_of(reg, ucc_tl_sharp_rcache_region_t, reg);
+        ucc_rcache_region_put(ctx->rcache, &region->super);
+    } else {
+        sharp_coll_dereg_mr(ctx->sharp_context, reg->mr);
+        ucc_free(reg);
+    }
+
+    return UCC_OK;
+}
+
+ucc_status_t ucc_tl_sharp_collective_progress(ucc_coll_task_t *coll_task)
+{
+    ucc_tl_sharp_task_t *task  = ucc_derived_of(coll_task, ucc_tl_sharp_task_t);
+    int completed;
+
+    if (task->req_handle != NULL) {
+        completed = sharp_coll_req_test(task->req_handle);
+        if (completed) {
+            if (task->super.args.coll_type == UCC_COLL_TYPE_ALLREDUCE) {
+                ucc_tl_sharp_mem_deregister(TASK_CTX(task), task->allreduce.s_mem_h);
+                ucc_tl_sharp_mem_deregister(TASK_CTX(task), task->allreduce.r_mem_h);
+            }
+
+            sharp_coll_req_free(task->req_handle);
+            coll_task->super.status = UCC_OK;
+        }
+    }
+
+    return coll_task->super.status;
+}
+
+ucc_status_t ucc_tl_sharp_barrier_start(ucc_coll_task_t *coll_task)
+{
+    ucc_tl_sharp_task_t *task  = ucc_derived_of(coll_task, ucc_tl_sharp_task_t);
+    ucc_tl_sharp_team_t *team  = TASK_TEAM(task);
+    int ret;
+
+    task->super.super.status = UCC_INPROGRESS;
+
+    ret = sharp_coll_do_barrier_nb(team->sharp_comm, &task->req_handle);
+    if (ret == SHARP_COLL_ENOT_SUPP || ret == SHARP_COLL_ENO_RESOURCE) {
+        return UCC_ERR_NO_RESOURCE;
```

Comment:
will need to extend #297 with fallback from "start". Currently in that PR i only made a fallback for the "init".